sign/bls/bls_test.go

```diff
@@ -6,23 +6,49 @@ import (
 
 	"github.com/stretchr/testify/require"
 	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/pairing"
+	"go.dedis.ch/kyber/v3/pairing/bls12381"
 	"go.dedis.ch/kyber/v3/pairing/bn256"
 	"go.dedis.ch/kyber/v3/util/random"
 )
 
-func TestBLS(t *testing.T) {
-	msg := []byte("Hello Boneh-Lynn-Shacham")
+func TestBn256(t *testing.T) {
 	suite := bn256.NewSuite()
+	testBLS(t, suite)
+	testBLSFailSig(t, suite)
+	testBLSFailKey(t, suite)
+	testBLSAggregateSignatures(t, suite)
+	testBLSFailAggregatedSig(t, suite)
+	testBLSFailAggregatedKey(t, suite)
+	testBLSBatchVerify(t, suite)
+	testBLSFailBatchVerify(t, suite)
+	testBinaryMarshalAfterAggregationIssue400(t, suite)
+}
+
+func TestBLS12381(t *testing.T) {
+	suite := bls12381.NewSuite()
+	testBLS(t, suite)
+	testBLSFailSig(t, suite)
+	testBLSFailKey(t, suite)
+	testBLSAggregateSignatures(t, suite)
+	testBLSFailAggregatedSig(t, suite)
+	testBLSFailAggregatedKey(t, suite)
+	testBLSBatchVerify(t, suite)
+	testBLSFailBatchVerify(t, suite)
+	testBinaryMarshalAfterAggregationIssue400(t, suite)
+}
+
+func testBLS(t *testing.T, suite pairing.Suite) {
+	msg := []byte("Hello Boneh-Lynn-Shacham")
 	private, public := NewKeyPair(suite, random.New())
```

Comment:
This comment is about a general design decision affecting the whole library, not just this Pull Request.

In BLS curves, G1 is small and fast and G2 is large and slow. Why? G1 are pairs of integers and G2 are pairs of complex integers, thus G2 takes twice the storage and requires more computation.
In practice, points are compressed for storage and transmission by dropping the y-coordinate, halving the amount of data. For BLS12-381, G1 points are halved from 96 bytes to 48 bytes, and G2 points are halved from 192 bytes to 96 bytes.

In this library, G1 is used for signatures and G2 is used for public keys: in other words, in dedis/kyber signatures require 48 bytes and public keys require 96 bytes.
This is the reverse of how BLS signatures are used in other blockchains, which swap your choice of G1 for G2. It's recommended in [draft-boneh-bls-signature-00](https://datatracker.ietf.org/doc/html/draft-boneh-bls-signature-00#section-2):

     2.  (minimizing public key size) Put public keys in G1 and signatures
       in G2.  This latter case comes up when we do signature
       aggregation, where most of the communication costs come from
       public keys.  **This is particularly relevant in applications such
       as blockchains** and compressing certificate chains, where the goal
       is to minimize the total size of multiple public keys and
       aggregated signatures.

and in a later edition of [draft-irtf-cfrg-bls-signature-04](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.1):

    2.  Minimal-pubkey-size: public keys are points in G1, signatures are
       points in G2.

       **Implementations using signature aggregation SHOULD use this
       approach, since the size of (PK_1, ..., PK_n, signature) is
       dominated by the public keys even for small n.**

For example, in Ethereum 2.0 they choose to minimise the public key size as recommended.

Reply:
This affects not only this Pull Request, but also ByzCoin, skipchain, and dedis/dela.

Your blockchains will be faster and lighter if you swap G1 for G2.

Reply:
Just want to point out that drand's fork (the one from which this work is derived) is modified to allow that flexibility; there's a general signing interface and BLS can be used on G1 or G2. See https://github.com/drand/kyber/blob/master/sign/bls/bls.go#L37 for example. (same for threshold BLS).